core/src/main/java/org/apache/accumulo/core/fate/ZooStore.java

```diff
@@ -91,6 +94,27 @@ public long create() {
     }
   }
 
+  @Override
+  protected void create(long tid, byte[] key) {
+    // TODO: Should we somehow make this Atomic or clean up on failure to make sure
+    // that either both of these writes happen or none happen?
+    try {
+      zk.putPersistentData(getTXPath(tid), TStatus.NEW.name().getBytes(UTF_8),
+          NodeExistsPolicy.FAIL);
+      // The key was already used to generate the tid but we still need to store it
+      // separate to check later for collision detection
+      zk.putPersistentData(getTXPath(tid) + "/" + TxInfo.TX_KEY, serializeTxInfo(key),
```

Comment:
This comment is not really for changes in this PR, just thinking ahead.  Eventually we will want to store the key in the ZK node at `getTXPath(tid)`.  This is because can only atomically modify a single node in ZK.  So eventually we can atomically set the key in ZK and Accumulo using conditional mutations in both.

In #3964 I did something like this by creating a [new struct](https://github.com/apache/accumulo/blob/208e39927310f8ff7dd5978f2794e26bee6be1b4/core/src/main/java/org/apache/accumulo/core/fate/ZooStore.java#L89) that is the value of the ZK node at `getTXPath(tid)` and then Accumulo's util functions for atomically modifying can be used like was done [here](https://github.com/apache/accumulo/blob/208e39927310f8ff7dd5978f2794e26bee6be1b4/core/src/main/java/org/apache/accumulo/core/fate/ZooStore.java#L227).

So I think for this key and #4131 we will need a new struct in that ZK node.

Changes could be made in this PR for this, but don't have to be made.

Reply:
This is now atomic in the latest updats